File: apps/worker/worker/documents/vocabulary.py
"""Corpus-backed skill vocabulary with deterministic phrase matching."""

import json
import re
from functools import lru_cache
from pathlib import Path
from typing import cast

CORPUS_PATH = Path(__file__).with_name("skills_corpus.json")

# Tokens keep internal punctuation used by skill names: c++, c#, node.js,
# ci/cd, t-sql. Dot/slash/hyphen runs must continue into alphanumerics so
# sentence punctuation stays outside; plus/hash runs may end a token.
TOKEN_PATTERN = re.compile(r"[a-z0-9]+(?:[.#/-]+[a-z0-9]+|[+#]+)*")
MAX_PHRASE_TOKENS = 6


class SkillVocabulary:
	def __init__(self, phrases: dict[str, str], categories: dict[str, str | None]) -> None:
		self.phrase_to_canonical = phrases
		self.categories = categories
		self.max_tokens = min(
			max(len(phrase.split()) for phrase in phrases),
			MAX_PHRASE_TOKENS,
		)

	def mention(self, text: str) -> dict[str, list[str]]:
		"""Match skill phrases longest-first. Returns canonical name -> matched phrases."""
		tokens = TOKEN_PATTERN.findall(text.casefold())
		found: dict[str, list[str]] = {}
		index = 0
		while index < len(tokens):
			match_count = 0
			for count in range(min(self.max_tokens, len(tokens) - index), 0, -1):
				phrase = " ".join(tokens[index : index + count])
				canonical = self.phrase_to_canonical.get(phrase)
				if canonical is not None:
					spans = found.setdefault(canonical, [])
					if phrase not in spans:
						spans.append(phrase)
					match_count = count
					break
			index += match_count or 1
		return found

	def category_for(self, canonical_name: str) -> str | None:
		return self.categories.get(canonical_name.casefold())


@lru_cache(maxsize=1)
def load_vocabulary() -> SkillVocabulary:
	corpus = cast(dict[str, object], json.loads(CORPUS_PATH.read_text()))
	raw_skills = cast(list[dict[str, object]], corpus["skills"])
	raw_aliases = cast(dict[str, str], corpus["aliases"])
	categories: dict[str, str | None] = {}
	phrases: dict[str, str] = {}
	for skill in raw_skills:
		name = str(skill["name"])
		category = skill.get("category")
		key = name.casefold()
		categories[key] = str(category) if category is not None else None
		phrases[key] = name
	for alias, canonical in raw_aliases.items():
		if canonical.casefold() not in categories:
			raise ValueError(f"Alias {alias!r} targets unknown skill {canonical!r}")
		phrases[alias.casefold()] = canonical
	return SkillVocabulary(phrases, categories)


def mentioned_skills(text: str) -> set[str]:
	"""Canonical skill names mentioned in free text such as job requirements."""
	return set(load_vocabulary().mention(text))

File: apps/worker/src/domain/vocabulary.ts
import corpus from "./skills_corpus.json";

export const TOKEN_PATTERN = /[a-z0-9]+(?:[.#/-]+[a-z0-9]+|[+#]+)*/g;
export const MAX_PHRASE_TOKENS = 6;

interface Corpus {
	readonly skills: readonly { readonly name: string; readonly category?: string | null }[];
	readonly aliases: Readonly<Record<string, string>>;
	readonly absorbers?: readonly string[];
}

export class SkillVocabulary {
	readonly phraseToCanonical: ReadonlyMap<string, string>;
	readonly categories: ReadonlyMap<string, string | null>;
	private readonly absorbers: ReadonlySet<string>;
	private readonly maxTokens: number;

	constructor(phrases: ReadonlyMap<string, string>, categories: ReadonlyMap<string, string | null>, absorbers: ReadonlySet<string>) {
		this.phraseToCanonical = phrases;
		this.categories = categories;
		this.absorbers = absorbers;
		this.maxTokens = Math.min(Math.max(...[...phrases.keys()].map((phrase) => phrase.split(" ").length)), MAX_PHRASE_TOKENS);
	}

	mention(text: string): Readonly<Record<string, readonly string[]>> {
		const tokens = tokenize(text);
		const found = new Map<string, string[]>();
		let index = 0;
		while (index < tokens.length) {
			let consumed = 0;
			for (let count = Math.min(this.maxTokens, tokens.length - index); count > 0; count -= 1) {
				const phrase = tokens.slice(index, index + count).join(" ");
				const canonical = this.phraseToCanonical.get(phrase);
				if (canonical !== undefined) {
					const spans = found.get(canonical) ?? [];
					if (!spans.includes(phrase)) spans.push(phrase);
					found.set(canonical, spans);
					consumed = count;
					break;
				}
				if (this.absorbers.has(phrase)) {
					consumed = count;
					break;
				}
			}
			index += consumed || 1;
		}
		return Object.fromEntries(found);
	}

	categoryFor(canonicalName: string): string | null {
		return this.categories.get(canonicalName.toLowerCase()) ?? null;
	}
}

let cached: SkillVocabulary | undefined;
export function loadVocabulary(): SkillVocabulary {
	if (cached) return cached;
	const data = corpus as Corpus;
	const categories = new Map<string, string | null>();
	const phrases = new Map<string, string>();
	for (const skill of data.skills) {
		const key = skill.name.toLowerCase();
		categories.set(key, skill.category ?? null);
		phrases.set(key, skill.name);
	}
	for (const [alias, canonical] of Object.entries(data.aliases)) {
		if (!categories.has(canonical.toLowerCase())) throw new Error(`Alias targets unknown skill: ${canonical}`);
		phrases.set(alias.toLowerCase(), canonical);
	}
	cached = new SkillVocabulary(phrases, categories, new Set((data.absorbers ?? []).map((item) => item.toLowerCase())));
	return cached;
}

export function mentionedSkills(text: string): Set<string> {
	return new Set(Object.keys(loadVocabulary().mention(text)));
}

export function tokenize(text: string): string[] {
	return text.toLowerCase().match(TOKEN_PATTERN) ?? [];
}
